src/backend/expression/case_expression.h

```diff
@@ -18,60 +20,55 @@ namespace expression {
-                 AbstractExpression *default_result)
+  // the first part is condition, and the second part is expression
+  typedef std::unique_ptr<AbstractExpression> AbstractExprPtr;
+  typedef std::pair<AbstractExprPtr, std::unique_ptr<AbstractExpression>>
+      WhenClause;
+  typedef std::unique_ptr<WhenClause> WhenClausePtr;
+  CaseExpression(ValueType vt, std::vector<WhenClausePtr> &t_clauses,
+                 WhenClausePtr default_clause)
       : AbstractExpression(EXPRESSION_TYPE_OPERATOR_CASE_EXPR),
-        clauses(clauses),
-        default_result(default_result),
-        case_type(vt) {}
-
-  ~CaseExpression() {
-    for (auto clause : clauses) {
-      delete clause;
-      clause = nullptr;
-    }
-    if (default_result != nullptr) {
-      delete default_result;
-      default_result = nullptr;
+        default_clause(std::move(default_clause)),
+        case_type(vt) {
+    for (auto &clause : t_clauses) {
+      clauses.push_back(std::move(clause));
     }
   }
 
   Value Evaluate(const AbstractTuple *tuple1, const AbstractTuple *tuple2,
                  executor::ExecutorContext *context) const {
-    for (auto clause : clauses) {
-      try {
-        return clause->Evaluate(tuple1, tuple2, context);
-      } catch (int ExceptionCode) {
-        if (ExceptionCode == 0) continue;
-        throw Exception("Unknown Exception code in CaseExpression.");
-      }
+    for (auto &clause : clauses) {
+      auto condition = clause->first->Evaluate(tuple1, tuple2, context);
+      if (condition.IsTrue())
+        return clause->second->Evaluate(tuple1, tuple2, context);
     }
-    return default_result->Evaluate(tuple1, tuple2, context);
+    return default_clause->second->Evaluate(tuple1, tuple2, context);
   }
 
   std::string DebugInfo(const std::string &spacer) const {
     return spacer + "CaseExpression";
   }
 
   AbstractExpression *Copy() const {
-    std::vector<AbstractExpression *> copied_clauses;
-    for (AbstractExpression *clause : clauses) {
-      if (clause == nullptr) {
-        continue;
-      }
-      copied_clauses.push_back(clause->Copy());
+    std::vector<WhenClausePtr> copied_clauses;
+    for (auto &clause : clauses) {
+      copied_clauses.push_back(WhenClausePtr(
+          new WhenClause(AbstractExprPtr(clause->first->Copy()),
+                         AbstractExprPtr(clause->second->Copy()))));
     }
-
-    return new CaseExpression(case_type, copied_clauses,
-                              default_result->Copy());
+    // default result has no condition.
+    return new CaseExpression(
+        case_type, copied_clauses,
+        WhenClausePtr(new WhenClause(
+            nullptr, AbstractExprPtr(default_clause->second->Copy()))));
   }
 
  private:
   // Case expression clauses
-  std::vector<AbstractExpression *> clauses;
+  std::vector<WhenClausePtr> clauses;
 
   // Fallback case result expression
-  AbstractExpression *default_result;
+  WhenClausePtr default_clause;
```

Comment:
Use the member variable naming convention by putting an underscore as a suffix. Do `clauses_` as an example.